Add integrations docs for Elixir (#9546)
getsentry/sentry-docs: docs/platforms/elixir/integrations/plug_and_phoenix/index.mdx

---
title: Plug and Phoenix
description: "Learn how to capture errors and their context in Plug pipelines and Phoenix applications."
sidebar_order: 10
---

> You can capture errors in [Plug](https://github.com/elixir-plug/plug) and [Phoenix](https://phoenixframework.org) applications with `Sentry.PlugContext` and `Sentry.PlugCapture`. `Sentry.PlugContext` adds contextual metadata from the current request, which is then included in errors that are captured and reported by `Sentry.PlugCapture`.

Reviewer: ```suggestion
You can capture errors in [Plug](https://github.com/elixir-plug/plug) and [Phoenix](https://phoenixframework.org) applications with `Sentry.PlugContext` and `Sentry.PlugCapture`. 

`Sentry.PlugContext` adds contextual metadata from the current request, which is then included in errors that are captured and reported by `Sentry.PlugCapture`.
```

Visually, it was hard to tell that this was two separate sentences so let's just break it up so it's easier to read :)

## Phoenix Configuration

> If you are using Phoenix:

Reviewer: ```suggestion
If you are using Phoenix, follow these steps to add the Plug integration:
```

  1. Add `Sentry.PlugCapture` above the `use Phoenix.Endpoint` line in your endpoint file.
  1. Add `Sentry.PlugContext` below `Plug.Parsers`.

> ```diff

Reviewer: ```suggestion
```Elixir
```
Was there a reason to say diff instead of elixir for the code block?

Author: This was originally in the docs for the SDK (on hex.pm), and I left it mostly untouched. I really do prefer the diff since it shows only the lines you need to add and where, while using Elixir syntax highlighting makes it harder to understand where the changes are IMO.

 defmodule MyAppWeb.Endpoint
+  use Sentry.PlugCapture
   use Phoenix.Endpoint, otp_app: :my_app

   # ...

   plug Plug.Parsers,
     parsers: [:urlencoded, :multipart, :json],
     pass: ["*/*"],
     json_decoder: Phoenix.json_library()

+  plug Sentry.PlugContext
```

### Capturing User Feedback

> If you would like to capture user feedback as described <PlatformLink to="/enriching-events/user-feedback/">here</PlatformLink>, the `Sentry.get_last_event_id_and_source/0` function can be used to see if Sentry has sent an event within the current Plug process (and get the source of that event). `:plug` will be the source for events coming from `Sentry.PlugCapture`. The options described in the Sentry documentation linked above can be encoded into the response as well.

Reviewer: ```suggestion
If you would like to <PlatformLink to="/enriching-events/user-feedback/">capture user feedback</PlatformLink>, the `Sentry.get_last_event_id_and_source/0` function can be used to see if Sentry has sent an event within the current Plug process (and get the source of that event). `:plug` will be the source for events coming from `Sentry.PlugCapture`. The options described in the Sentry documentation linked above can be encoded into the response as well.
```

> An example Phoenix application setup that displays the user feedback form on 500 responses on requests accepting HTML could look like this:

Reviewer: ```suggestion
Below is an example Phoenix application setup that displays the user feedback form on `500` responses on requests accepting HTML:
```

```elixir
defmodule MyAppWeb.ErrorView do
  # ...

  def render("500.html", _assigns) do
    case Sentry.get_last_event_id_and_source() do
      {event_id, :plug} when is_binary(event_id) ->
        opts = Jason.encode!(%{eventId: event_id})

        ~E"""
        <script src="https://browser.sentry-cdn.com/5.9.1/bundle.min.js" integrity="sha384-/x1aHz0nKRd6zVUazsV6CbQvjJvr6zQL2CHbQZf3yoLkezyEtZUpqUNnOLW9Nt3v" crossorigin="anonymous"></script>
        <script>
          Sentry.init({ dsn: '<%= Sentry.Config.dsn() %>' });
          Sentry.showReportDialog(<%= raw opts %>)
        </script>
        """

      _ ->
        "Error"
    end
  end
end
```

## Plug Configuration

> If you are in a non-Phoenix Plug application:

Reviewer: ```suggestion
If you have a non-Phoenix application, follow these steps to configure the Plug integration:
```

>   1. Add `Sentry.PlugCapture` at the top of your Plug application
>   1. Add `Sentry.PlugContext` below `Plug.Parsers` (if it is in your stack)

Reviewer: ```suggestion
  1. Add `Sentry.PlugCapture` at the top of your Plug application.
  1. Add `Sentry.PlugContext` below `Plug.Parsers` (if it is in your stack).
```

> ```diff

Reviewer: ```suggestion
```Elixir
```

Author: Same as the other comment 🙃
